### ISSUE TITLE
ForkJoinPoolStarvationSpec

### ISSUE THREAD

**Issue by He-Pin** [apache/incubator-pekko#661]
```
- must not starve tasks arriving from external dispatchers under high internal traffic *** FAILED *** (6 seconds, 14 milliseconds)
[09-20 10:18:16.027] [info]   java.lang.AssertionError: assertion failed: timeout (6 seconds) during expectMsg while waiting for All fine
[09-20 10:18:16.027] [info]   at scala.Predef$.assert(Predef.scala:223)
[09-20 10:18:16.027] [info]   at org.apache.pekko.testkit.TestKitBase.expectMsg_internal(TestKit.scala:472)
[09-20 10:18:16.028] [info]   at org.apache.pekko.testkit.TestKitBase.expectMsg(TestKit.scala:449)
[09-20 10:18:16.028] [info]   at org.apache.pekko.testkit.TestKitBase.expectMsg$(TestKit.scala:449)
[09-20 10:18:16.028] [info]   at org.apache.pekko.testkit.TestKit.expectMsg(TestKit.scala:982)
[09-20 10:18:16.028] [info]   at org.apache.pekko.dispatch.ForkJoinPoolStarvationSpec.$anonfun$new$6(ForkJoinPoolStarvationSpec.scala:78)
[09-20 10:18:16.028] [info]   at org.apache.pekko.dispatch.ForkJoinPoolStarvationSpec.$anonfun$new$6$adapted(ForkJoinPoolStarvationSpec.scala:75)
[09-20 10:18:16.028] [info]   at scala.collection.immutable.Range.foreach(Range.scala:158)
[09-20 10:18:16.028] [info]   at org.apache.pekko.dispatch.ForkJoinPoolStarvationSpec.$anonfun$new$2(ForkJoinPoolStarvationSpec.scala:75)
[09-20 10:18:16.029] [info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[09-20 10:18:16.029] [info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[09-20 10:18:16.029] [info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[09-20 10:18:16.029] [info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[09-20 10:18:16.029] [info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[09-20 10:18:16.029] [info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[09-20 10:18:16.029] [info]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1239)
[09-20 10:18:16.029] [info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[09-20 10:18:16.030] [info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[09-20 10:18:16.030] [info]   at org.apache.pekko.testkit.PekkoSpec.withFixture(PekkoSpec.scala:80)
[09-20 10:18:16.030] [info]   at org.scalatest.wordspec.AnyWordSpecLike.invokeWithFixture$1(AnyWordSpecLike.scala:1237)
[09-20 10:18:16.030] [info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTest$1(AnyWordSpecLike.scala:1249)
[09-20 10:18:16.030] [info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[09-20 10:18:16.030] [info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest(AnyWordSpecLike.scala:1249)
[09-20 10:18:16.030] [info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest$(AnyWordSpecLike.scala:1231)
[09-20 10:18:16.031] [info]   at org.apache.pekko.testkit.PekkoSpec.runTest(PekkoSpec.scala:80)
[09-20 10:18:16.031] [info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTests$1(AnyWordSpecLike.scala:1308)
[09-20 10:18:16.031] [info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[09-20 10:18:16.031] [info]   at scala.collection.immutable.List.foreach(List.scala:431)
[09-20 10:18:16.031] [info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[09-20 10:18:16.031] [info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:390)
[09-20 10:18:16.031] [info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:427)
[09-20 10:18:16.031] [info]   at scala.collection.immutable.List.foreach(List.scala:431)
[09-20 10:18:16.032] [info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[09-20 10:18:16.032] [info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[09-20 10:18:16.032] [info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[09-20 10:18:16.032] [info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests(AnyWordSpecLike.scala:1308)
[09-20 10:18:16.032] [info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests$(AnyWordSpecLike.scala:1307)
[09-20 10:18:16.032] [info]   at org.apache.pekko.testkit.PekkoSpec.runTests(PekkoSpec.scala:80)
[09-20 10:18:16.032] [info]   at org.scalatest.Suite.run(Suite.scala:1114)
[09-20 10:18:16.032] [info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[09-20 10:18:16.033] [info]   at org.apache.pekko.testkit.PekkoSpec.org$scalatest$wordspec$AnyWordSpecLike$$super$run(PekkoSpec.scala:80)
[09-20 10:18:16.033] [info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$run$1(AnyWordSpecLike.scala:1353)
[09-20 10:18:16.033] [info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[09-20 10:18:16.033] [info]   at org.scalatest.wordspec.AnyWordSpecLike.run(AnyWordSpecLike.scala:1353)
[09-20 10:18:16.033] [info]   at org.scalatest.wordspec.AnyWordSpecLike.run$(AnyWordSpecLike.scala:1351)
[09-20 10:18:16.033] [info]   at org.apache.pekko.testkit.PekkoSpec.org$scalatest$BeforeAndAfterAll$$super$run(PekkoSpec.scala:80)
[09-20 10:18:16.033] [info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[09-20 10:18:16.033] [info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[09-20 10:18:16.034] [info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[09-20 10:18:16.034] [info]   at org.apache.pekko.testkit.PekkoSpec.run(PekkoSpec.scala:80)
[09-20 10:18:16.034] [info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[09-20 10:18:16.034] [info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[09-20 10:18:16.034] [info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
[09-20 10:18:16.034] [info]   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[09-20 10:18:16.034] [info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[09-20 10:18:16.034] [info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[09-20 10:18:16.035] [info]   at java.base/java.lang.Thread.run(Thread.java:1583)
```

refs: https://github.com/apache/incubator-pekko/actions/runs/6247109120/job/16959068675#step:6:3238

**Comment by He-Pin** [comment]
[rs](https://github.com/apache/incubator-pekko/actions/runs/6247109120/job/16959069348#step:6:3153)

**Comment by He-Pin** [comment]
refs: https://github.com/openjdk/jdk/pull/11319

**Comment by He-Pin** [comment]
@jrudolph You was doing some design on this, would you like to work on this to call `externalSubmit` etc thing after some iteration?

**Comment by He-Pin** [comment]
Status: this test is already skipped on JDK 21 (`e94e7b971b` =test Skip ForkJoinPoolStarvationSpec on JDK 21) and has open POC PRs (#665 closed, #666 open) exploring `externalSubmit`-based yielding.

The failure ("must not starve tasks arriving from external dispatchers under high internal traffic" timing out waiting for "All fine") reflects ForkJoinPool fairness/starvation behavior across JDKs, not a test-timeout bug. A blind timeout bump would mask the starvation it is meant to detect. Recommend tracking via the POC PRs rather than a test-only tweak; closing/keeping open should follow that effort.

**Comment by He-Pin** [comment]
Investigation update (2026-07-11): I re-tested the old `externalSubmit` yielding POC from `fork/fjyield` (`8cad2b196d`) against JDK 21 and JDK 25.

Results:

- JDK 21.0.5: `ForkJoinPoolStarvationSpec` passed (1 test, 1000 ping/reply iterations, 46 ms).
- JDK 25.0.2: the same POC failed twice, timing out after about 3 seconds waiting for `All fine`.
- Instrumentation on JDK 25 confirmed that the POC really called `ForkJoinPool.externalSubmit`: the pool's submission count increased after each throughput-boundary resubmission and the steal count kept increasing. This rules out reflection/version detection or a fallback to ordinary `execute` as the cause.

The JDK 25 failure is a second fairness problem. `externalSubmit` moves the resubmitted mailbox from a worker-local queue to a shared submission queue, but it does not guarantee fairness between different submission queues. With two continuously replenished queues and parallelism 2, each worker can remain attached to one busy queue indefinitely while the innocent actor's submission queue is never visited.

This matches the still-unresolved OpenJDK report [JDK-8373224](https://bugs.openjdk.org/browse/JDK-8373224), which describes workers continuously stealing from replenished external queues while other external queues starve. The original local-vs-external starvation regression is [JDK-8315740](https://bugs.openjdk.org/browse/JDK-8315740), resolved as `Won't Fix`. `externalSubmit` was added in JDK 20 and only guarantees placement into a submission queue; it does not specify cross-queue fairness ([JDK-8297648](https://bugs.openjdk.org/browse/JDK-8297648)).

Conclusion:

- The POC can make this exact test pass on JDK 21.0.5, but that is not a portable fairness guarantee.
- On JDK 25.0.2, `externalSubmit` alone is not sufficient.
- JDK 17-19 do not expose `externalSubmit`; relying on reflective access to private `externalPush` plus `--add-opens` would not be an acceptable library/runtime contract.
- Keeping this spec pending on JDK 17+ is currently the safest option if Pekko retains the existing `ForkJoinPool` architecture.

Removing the skip robustly would require an architectural change under Pekko's control, such as a centralized external handoff queue/thread for throughput-boundary yields, or using a fairer executor. That would change a hot scheduling path and needs benchmarks, shutdown/race coverage, and broader dispatcher regression testing; it should not be treated as a test-timeout fix.

**Comment by He-Pin** [comment]
As This seems can not be fixed totally on our side, we can leave as it is.
